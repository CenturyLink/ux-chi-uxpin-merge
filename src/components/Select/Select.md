# Select

```javascript
<Select width={300}>
    <SelectItem>Option 1</SelectItem>
    <SelectItem>Option 2</SelectItem>
    <SelectItem>Option 3</SelectItem>
    <SelectItem>Option 4</SelectItem>
    <SelectItem>Option 5</SelectItem>
    <SelectItem>Option 5 Option 5 Option 5 Option 5 Option 5</SelectItem>
</Select>
```