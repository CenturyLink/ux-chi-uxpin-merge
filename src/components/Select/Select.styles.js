import styled from '@emotion/styled';
import colors from '../../styles/colors';
import shadows from '../../styles/shadows';
import sizes from '../../styles/sizes';
import typography from '../../styles/typography';

const getWidth = (width) => {
  if (width === 'stretched') {
    return '100%';
  }

  if (width === 'auto') {
    return 'auto';
  }

  return `${width}px`;
};

const SelectContainerStyles = styled('div')`
    width: ${(props) => (props.stretched ? '100%' : 'auto')};
    font-family: ${typography.fontFamily};
    font-weight: normal;
    font-size: ${typography.size.l}px;
    line-height: ${sizes.m * typography.lineHeightMultiplier}px;
    color: ${colors.gray.base};
`;

const SelectStyles = styled('div')`
    display: inline-flex;
    flex: 1 0 auto;
    align-items: center;
    justify-content: space-between;
    width: ${(props) => getWidth(props.width)};
    min-width: 250px;
    height: ${sizes.xxl}px;
    margin-bottom: 3px;
    border-bottom: 1px solid ${colors.silver.dark20};

    button {
      margin-left: auto;
    }
    svg {
      min-width: fit-content;
    }
`;

const SelectContent = styled('div')`
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    cursor: pointer;
`;

const MenuStyles = styled('ul')`
    position: absolute;
    display: flex;
    flex-direction: column;
    left: 0;
    width: ${(props) => getWidth(props.width)};
    min-width: 250px;
    margin: 1px 0;
    padding: ${sizes.s}px 0 ${sizes.s}px 0;
    background: white;
    box-shadow: ${shadows.box};
    list-style: none;
`;

const ItemStyles = styled('li')`
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    justify-content: flex-start;
    position: relative;
    width: inherit;
    height: ${sizes.xxl}px;
    padding: 0 ${sizes.m}px;
    background-color: ${(props) => (props.id === `${props.active}listItem` ? colors.silver.light1 : 'white')};
    cursor: pointer;
    &:hover {
        background-color: ${colors.silver.light1};
    }
    span {
        max-width: 90%;
        overflow: hidden;
        white-space: nowrap;
        text-align: center;
        text-overflow: ellipsis;
    }
`;

export {
  SelectStyles, MenuStyles, ItemStyles, SelectContainerStyles, SelectContent,
};
