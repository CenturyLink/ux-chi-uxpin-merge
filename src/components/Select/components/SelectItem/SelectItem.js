import PropTypes from 'prop-types';
import React from 'react';

/**
 * @uxpinnamespace Select
 */
export default function SelectItem(props) {
  return (
    <span>
      {props.children}
    </span>
  );
}

SelectItem.propTypes = {
  /** @uxpinpropname Label */
  children: PropTypes.string,
};
