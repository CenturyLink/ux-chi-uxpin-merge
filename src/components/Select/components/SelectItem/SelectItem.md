# SelectItem

```javascript
<SelectItem>Option 1</SelectItem>
```