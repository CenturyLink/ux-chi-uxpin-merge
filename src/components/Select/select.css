select.chi-select {
    background-image: url("data:image/svg+xml;charset=utf-8,%3Csvg width='16' height='16' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M8.097 10.66l6.005-6.216.58-.6 1.162 1.201-.58.601-6.568 6.798-.58.601-.648-.67-.069-.067L.601 5.741 0 5.161 1.202 4l.6.58 6.295 6.08z' fill='%230262B9' fill-rule='evenodd'/%3E%3C/svg%3E") !important;
    background-position: right .5rem center !important;
    background-repeat: no-repeat !important;
}