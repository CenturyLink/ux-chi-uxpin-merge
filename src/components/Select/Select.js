import PropTypes from 'prop-types';
import React from 'react';
import Button from '../Button/Button';
import Icon from '../Icon/Icon';
import SelectItem from './components/SelectItem/SelectItem';
import {
  ItemStyles,
  MenuStyles,
  SelectContainerStyles,
  SelectContent,
  SelectStyles,
} from './Select.styles';

export default class Select extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      content: this.props.placeholder,
      open: this.props.defaultOpen,
    };
    this.toggleList = this.toggleList.bind(this);
  }

  selectItem(content, id) {
    this.setState({
      content,
      open: false,
      selected: id,
    });
  }

  toggleList() {
    this.setState((prevState) => ({
      open: !prevState.open,
    }));
  }

  restartSelect(e) {
    e.stopPropagation();
    this.setState({ content: this.props.placeholder, open: false, selected: '' });
  }

  render() {
    return (
      <SelectContainerStyles {...this.props}>
        <SelectStyles onClick={this.toggleList} {...this.props}>
          <SelectContent>
            {this.state.content}
          </SelectContent>
          {this.state.content !== this.props.placeholder
            ? (
              <Button
                mode="minimal"
                type="error"
                size="s"
                stretched={false}
                icon={<Icon icon="ActionCloseSvg" size="xs" color="#444" />}
                onClick={(e) => {
                  this.restartSelect(e);
                  this.props.onRestart();
                }}
              />
            )
            : ''}
          <Icon icon="ChevronSvg" color="#444" />
        </SelectStyles>
        {this.state.open
          ? (
            <MenuStyles {...this.props}>
              {React.Children.map(this.props.children, (item, i) => {
                const option = item.props.children;
                return (
                  <ItemStyles
                    id={`${i}listItem`}
                    active={this.state.selected}
                    onClick={() => {
                      this.selectItem(option, i);
                      this.props.onSelect();
                    }}
                    {...this.props}>
                    {item}
                  </ItemStyles>
                );
              })}
            </MenuStyles>
          ) : ''}
      </SelectContainerStyles>
    );
  }
}

Select.SelectItem = SelectItem;

Select.propTypes = {
  children: PropTypes.node,
  defaultOpen: PropTypes.bool,
  /** @uxpinignoreprop */
  onRestart: PropTypes.func,
  onSelect: PropTypes.func,
  placeholder: PropTypes.string,
  width: PropTypes.oneOf(['stretched', 'auto']),
};

Select.defaultProps = {
  defaultOpen: false,
  placeholder: 'Select...',
  width: 'stretched',
};
