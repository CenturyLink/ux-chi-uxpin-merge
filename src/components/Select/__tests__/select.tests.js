import React from "react";
import * as enzyme from "enzyme";
import toJson from "enzyme-to-json";
import SelectItem from "../components/SelectItem/SelectItem";
import Select from "../Select";

test("renders Select with width stretched correctly", () => {
    const select = (
      <Select width="stretched">
        <SelectItem>Option 1</SelectItem>
        <SelectItem>Option 2</SelectItem>
        <SelectItem>Option 3</SelectItem>
        <SelectItem>Option 4</SelectItem>
        <SelectItem>Option 5</SelectItem>
        <SelectItem>Option 5 Option 5 Option 5 Option 5 Option 5</SelectItem>
      </Select>
    );
  
    expect(toJson(enzyme.shallow(select))).toMatchSnapshot(`enzyme.shallow`);
  });

test("renders Select with width auto correctly", () => {
    const select = (
      <Select width="auto">
        <SelectItem>Option 1</SelectItem>
        <SelectItem>Option 2</SelectItem>
        <SelectItem>Option 3</SelectItem>
        <SelectItem>Option 4</SelectItem>
        <SelectItem>Option 5</SelectItem>
        <SelectItem>Option 5 Option 5 Option 5 Option 5 Option 5</SelectItem>
      </Select>
    );
  
    expect(toJson(enzyme.shallow(select))).toMatchSnapshot(`enzyme.shallow`);
  });
