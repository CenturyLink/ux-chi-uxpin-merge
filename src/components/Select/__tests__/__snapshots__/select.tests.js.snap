// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Select with width auto correctly: enzyme.shallow 1`] = `
<SelectContainerStyles
  defaultOpen={false}
  placeholder="Select..."
  width="auto"
>
  <SelectStyles
    defaultOpen={false}
    onClick={[Function]}
    placeholder="Select..."
    width="auto"
  >
    <SelectContent>
      Select...
    </SelectContent>
    <Icon
      color="#444"
      icon="ChevronSvg"
    />
  </SelectStyles>
</SelectContainerStyles>
`;

exports[`renders Select with width stretched correctly: enzyme.shallow 1`] = `
<SelectContainerStyles
  defaultOpen={false}
  placeholder="Select..."
  width="stretched"
>
  <SelectStyles
    defaultOpen={false}
    onClick={[Function]}
    placeholder="Select..."
    width="stretched"
  >
    <SelectContent>
      Select...
    </SelectContent>
    <Icon
      color="#444"
      icon="ChevronSvg"
    />
  </SelectStyles>
</SelectContainerStyles>
`;
